### ISSUE TITLE
[cling] Crash unloading a templated function containing a lambda expression

### ISSUE THREAD

**Issue by michael-mpik** [root-project/root#9850]
// cling problem with lambda functions inside of template functions
//
// A code fragment illustrating the problem is given below.  
// Starting ROOT and running the macro once works, running it 
// a second time produces a crash. Commenting the (not-even used)
// lambda function avoid the crash. The macro can be run as often as
// one likes.
//
// ROOT 6.20/02  on MacOSX 10.15.06
// root [1] .x crash.cc    -> works
// root [2] .x crash.cc    -> crash

template<class T>  T square(T x) {
  auto crash = [](double x) {return x;};  // comment this line to avoid the crash
  return x*x;
}

void crash()
{
  printf("%d\n",square(2));
}

**Comment by vepadulano** [comment]
Works on lxplus with latest ROOT release 6.24.06 (6.26 is due soon)
```
[vpadulan@lxplus772 ~]$source /cvmfs/sft.cern.ch/lcg/views/LCG_101/x86_64-centos7-gcc11-opt/setup.sh
[vpadulan@lxplus772 ~]$root --version
ROOT Version: 6.24/06
Built for linuxx8664gcc on Sep 02 2021, 14:20:23
From tags/v6-24-06@v6-24-06
[vpadulan@lxplus772 ~]$cat crash.cc
#include <stdio.h>

template <typename T> auto square(T x) {
auto crash = [](double x) {return x;};
return x*x;
}

void crash()
{
printf("%d\n",square(2));
}
[vpadulan@lxplus772 ~]$root
   ------------------------------------------------------------------
  | Welcome to ROOT 6.24/06                        https://root.cern |
  | (c) 1995-2021, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Sep 02 2021, 14:20:23                 |
  | From tags/v6-24-06@v6-24-06                                      |
  | With g++ (GCC) 11.1.0                                            |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------

root [0] .x crash.cc
4
root [1] .x crash.cc
4
root [2] .x crash.cc
4
root [3] .x crash.cc
4
root [4] .x crash.cc
4
```

**Comment by vepadulano** [comment]
Four our reference, found the issue in ROOT 6.22.06 on lxplus with a debug build
```
[vpadulan@lxplus772 ~]$source /cvmfs/sft.cern.ch/lcg/views/LCG_99/x86_64-centos7-gcc10-dbg/setup.sh
[vpadulan@lxplus772 ~]$root --version
ROOT Version: 6.22/06
Built for linuxx8664gcc on Nov 27 2020, 15:14:08
From tags/v6-22-06@v6-22-06
[vpadulan@lxplus772 ~]$cat crash.cc
#include <stdio.h>

template <typename T> auto square(T x) {
auto crash = [](double x) {return x;};
return x*x;
}

void crash()
{
printf("%d\n",square(2));
}
[vpadulan@lxplus772 ~]$root
   ------------------------------------------------------------------
  | Welcome to ROOT 6.22/06                        https://root.cern |
  | (c) 1995-2020, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Nov 27 2020, 15:14:08                 |
  | From tags/v6-22-06@v6-22-06                                      |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------

root [0] .x crash.cc
4
root [1] .x crash.cc
root.exe: /build/jenkins/workspace/lcg_release_pipeline/build/projects/ROOT-v6.22.06/src/ROOT/v6.22.06/interpreter/llvm/src/tools/clang/lib/AST/DeclCXX.cpp:1106: clang::CXXMethodDecl* clang::CXXRecordDecl::getLambdaCallOperator() const: Assertion `!Calls.empty() && "Missing lambda call operator!"' failed.
```
We are probably missing some backport, to be decided if it is needed.
@michael-mpik  if possible please upgrade to the latest ROOT version and let us know if the problem persists. Thanks again for your report :smile:

**Comment by vepadulano** [comment]
I tried on one of our CI nodes with MacOS 10.15 at the tip of the 6.24 branch and still the problem is not shown.
```
macitois19:6-24 sftnight$ root
   ------------------------------------------------------------------
  | Welcome to ROOT 6.24/07                        https://root.cern |
  | (c) 1995-2021, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for macosx64 on Feb 09 2022, 23:19:43                      |
  | From remotes/origin/v6-24-00-patches@v6-24-06-41-gad88a28875     |
  | With Apple clang version 12.0.0 (clang-1200.0.32.28)             |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------

root [0] .x crash.cc
4
root [1] .x crash.cc
4
root [2] .x crash.cc
4
root [3] .x crash.cc
4
```

In your example above, `From ***@***.***` looks quite weird to me. To better understand the differences, it would be helpful to know how you installed ROOT.

**Comment by michael-mpik** [comment]
good morning!

I tried with root 6.24/06, and installed either as .pkg or from tar.gz - both with the same result. I will do some more tests, and in particular also try to upgrade Xcode11 to Xcode12, and let you know.

--
Michael Schmelling, MPI for Nuclear Physics
Phone:+49-6221-516-511 Fax:+49-6221-516-603

**Comment by michael-mpik** [comment]
I now installed Xcode 12.4 (12D4e) and root6.24/06 from root_v6.24.06.macos-10.15-x86_64-clang120.pkg. My system is macOS Catalina 10.15.6. Here is what I get:


root [0] .x crash.cc 
4
root [1] .x crash.cc 

 *** Break *** segmentation violation
[/usr/lib/system/libsystem_platform.dylib] _sigtramp (no debug info)
[<unknown binary>] (no debug info)
[/Applications/root_v6.24.06/lib/libCling.6.24.06.so] clang::RecursiveASTVisitor<cling::(anonymous namespace)::StaticVarCollector>::dataTraverseNode(clang::Stmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) (no debug info)
...

--
Michael Schmelling, MPI for Nuclear Physics
Phone:+49-6221-516-511 Fax:+49-6221-516-603

**Comment by jalopezg-r00t** [comment]
Thanks for reporting @michael-mpik.  I have reproduced the problem and a PR is on its way.

**Comment by jalopezg-r00t** [comment]
TLDR; for reference, the issue is described below.

Implicit instantiation of a function template calls `DeclCollector::HandleCXXImplicitFunctionInstantiation()`, which appends the FunctionDecl to the transaction.  According to the [docs](https://clang.llvm.org/doxygen/classclang_1_1ASTConsumer.html#a880e9a2fd04c8abd5cd218e0a4ed2e56), the body of the function has not yet been instantiated.  `HandleTopLevelDecl()` will be called again for this decl at the end of the TU, which will append it again to the transaction (same `Decl *`, different DCI).

This is by design.  However, unloading of decls in the transaction should _not_ process the same `Decl *` twice.  In the reported case, traversing the AST for a second time triggers an assertion failure in `clang::CXXRecordDecl::getLambdaCallOperator()`, as part of the members of the underlying CXXRecordDecl (of a lambda) are not there anymore.
